### DISCUSSION TITLE
TOSCA 2.0 editor? or valid JSON schema that can be used with YAML editor?

### DISCUSSION THREAD

**Discussion by cdanger** [oasis-open/tosca-community-contributions#168] [Q&A]
Hello, what are you guys using to edit TOSCA 2.0 files? Any TOSCA-2.0-aware editor out there?
Or as an alternative, has anybody written a JSON schema for TOSCA 2.0? 
This could be fed to a YAML editor such as the [RedHat VScode extension ](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml) for basic syntax validation at least.

**Comment by tliron** [comment]
I have begun work on a [VS code extension](https://github.com/tliron/puccini-vscode/), but work has stalled due to lack of time. :) If anyone would like to help out, welcome!

I don't think a JSON schema would be especially helpful, TOSCA is just too complex for simple syntax.

**Reply by koppor** [reply]
In the context of the Eclipse Foundation, I started a [Google Summer of Code](https://summerofcode.withgoogle.com/) [project](https://gitlab.eclipse.org/eclipsefdn/emo-team/gsoc-at-the-ef/-/issues/10) to add a language server for TOSCA 2.0. We found @omarawd7 as GSoC student and work as began. Since both Omar and me are more comfortable in the field of Java, we currently opted to start implementation in Java. We evaluated [lsp4j](https://github.com/eclipse-lsp4j/lsp4j) and [tosca-yaml-lsp](https://github.com/raghvendra-v/tosca-yaml-lsp) as a basis. We found that lsp4j is the lower risk. The [Xtext](https://eclipse.dev/Xtext/) framework is a too high entry barrier.

The project is in early stages. We could switch to another language. However, the result still needs to be under the roof of the Eclipse foundation. 😅

Looking forward where this trip will lead us to.

**Reply by lauwers** [reply]
Thanks @koppor and Omar. Very exciting. We look forward to hearing about your experience.

**Comment by cdanger** [comment]
I agree certainly it would not capture the full TOSCA semantics, but it is still better than nothing for basic syntax validation in the IDE (vscode or others), which is missing terribly in my case.

**Comment by pmjordan** [comment]
In the absence of anything else I used [puccini](https://github.com/tliron/puccini) to validate each time I saved and turned on autosave.

**Comment by koppor** [comment]
[Winery](https://github.com/winery/winery) should have full TOSCA 2.0 web-based modelling support. One part for defining types, the other part for crafting the application topology.

Currently, Winery also offers support for TOSCA 1.0 and research extensions. Therefore, it offers more functionality than required for pure modelling.

I am starting to work on cleaning it up and trying to move it to a TOSCA 2.0 only tooling.

---

Be aware that "Winery" is fully called "Eclipse Winery". "Eclipse" is a hint to the "Eclipse Foundation" be understood as open source foundation similar to the Apache Foundation. It implies strong backing of an open source foundation. It does not mean, it runs on the Eclipse IDE. Eclipse Winery is implemented in Angular and offers a web-based modeling experience.

**Reply by tliron** [reply]
Excellent! For what it's worth, though Puccini is written in Go, it does have [Java bindings](https://github.com/tliron/puccini/tree/main/wrappers/java) (that can be improved).

**Reply by lauwers** [reply]
Thanks @koppor. Could you elaborate on the *research extensions* in Winery?

**Comment by Shishqa** [comment]
Ironically, I've also started working on the TOSCA 1.3-2.0 language server (`totosca`): https://github.com/Shishqa/totosca.

It is in its early stage but can be handy even so. It can:
- validate TOSCA 1.3 and 2.0 schemas (currently, some parts are missing; working on reaching ~100% coverage)
- provide the goto-definition capability for type references and other definition references (artifacts, requirement targets, etc)
- find references for types (aka find usages)
- provide (hacky-buggy) suggestions for types (working on schema-level suggestions)

It is written in Rust, which has a powerful macro system helping leverage many entities in the standard. For YAML parsing, I use [yaml-peg-rs](https://github.com/KmolYuan/yaml-peg-rs) (without the `serde`), and [petgraph](https://github.com/petgraph/petgraph) is used for storing AST and handling the generic nature of relations between different TOSCA entities. And last but not least is [ariadne](https://github.com/zesterer/ariadne), which helps display fancy diagnostics messages.

![](https://github.com/Shishqa/totosca/blob/main/docs/assets/toto-check.png?raw=true)

You can check it out by installing it with cargo:
```
cargo install --git https://github.com/Shishqa/totosca toto_cli
```

Try with your template:
```
toto check your_template.yaml
```

It also has some editor integrations:
- [neovim](https://github.com/Shishqa/totosca/tree/main/integrations/nvim)
- [vscode](https://github.com/Shishqa/totosca/tree/main/integrations/vscode)

**Reply by tliron** [reply]
Nice! I've added it to the [wiki](https://github.com/oasis-open/tosca-community-contributions/wiki/Known-TOSCA-Implementations).

**Reply by lauwers** [reply]
Thanks @Shishqa. Very exciting. Feel free to provide feedback on which features make it difficult for you to reach 100% coverage.
